dali/pipeline/operator/builtin/external_source_test.cc

```diff
@@ -22,9 +24,189 @@
 #include "dali/pipeline/executor/async_pipelined_executor.h"
 #include "dali/pipeline/operator/builtin/external_source.h"
 #include "dali/test/dali_test_config.h"
+#include "dali/c_api.h"
 
 namespace dali {
 
+namespace {
+
+/**
+ * Load the test file as a binary string.
+ */
+std::string LoadTestFile(const std::string &test_file_path) {
+  std::ifstream fin(test_file_path, std::ios::binary);
+  if (!fin)
+    throw std::runtime_error(std::string("Failed to open file: ") + test_file_path);
+  std::stringstream ss;
+  ss << fin.rdbuf();
+  return ss.str();
+}
+
+}  // namespace
+
+template<typename Backend>
+class ExternalSourceBasicTest : public ::testing::Test {
+ protected:
+  static constexpr bool is_cpu = std::is_same_v<Backend, CPUBackend>;
+
+  void CreateAndSerializePipeline(bool repeat_last) {
+    auto pipeline = std::make_unique<Pipeline>(batch_size_, num_threads_, device_id_, false, 1,
+                                               false);
+    pipeline->AddOperator(
+            OpSpec("ExternalSource")
+                    .AddArg("repeat_last", repeat_last)
+                    .AddArg("device", std::string{is_cpu ? "cpu" : "gpu"})
```

Comment:
Move `is_cpu ? "cpu" : "gpu"` to a variable (you are using it several times)

Reply:
Done